react-common/components/controls/TeachingBubble.tsx

```diff
@@ -0,0 +1,291 @@
+import * as React from "react";
+import * as ReactDOM from "react-dom";
+import { Button } from "./Button";
+import { ContainerProps, classList } from "../util";
+import { FocusTrap } from "./FocusTrap";
+import { useEffect } from "react";
+
+export enum Location {
+    Above,
+    Below,
+    Left,
+    Right,
+    Center
+}
+
+export interface TargetContent {
+    title: string;
+    description: string;
+    targetQuery: string;
+    location: Location;
+}
+
+export interface TeachingBubbleProps extends ContainerProps {
+    targetContent: TargetContent;
+    stepNumber: number;
+    totalSteps: number;
+    onClose: () => void;
+    parentElement?: Element;
+    activeTarget?: boolean; // if true, the target is clickable
+    onNext: () => void;
+    onBack: () => void;
+}
+
+export const TeachingBubble = (props: TeachingBubbleProps) => {
+    const {
+        id,
+        className,
+        ariaLabel,
+        ariaHidden,
+        ariaDescribedBy,
+        role,
+        targetContent,
+        onClose,
+        onNext,
+        onBack,
+        stepNumber,
+        totalSteps,
+        parentElement,
+        activeTarget
+    } = props;
+
+    const margin = 10;
+    const transparentBorder = `${margin}px solid transparent`;
+    const opaqueBorder = `${margin}px solid`;
+    let bubble: HTMLElement;
+    let bubbleArrow: HTMLElement;
+    let bubbleBounds: DOMRect;
+    let targetElement: HTMLElement;
+    let targetBounds: DOMRect;
+    let top: number;
+    let left: number;
+    let arrowTop: number;
+    let arrowLeft: number;
+
+    useEffect(() => {
+        positionBubbleAndCutout();
+        window.addEventListener("resize", positionBubbleAndCutout);
+        return () => {
+            window.removeEventListener("resize", positionBubbleAndCutout);
+        }
+    }, [stepNumber]);
+
+    useEffect(() => {
+        if (!hasSteps) {
+            const footer = document.querySelector(".teaching-bubble-footer") as HTMLElement;
+            footer.style.flexDirection = "row-reverse";
```

Comment:
It's to change the style in the case that the tour consists of just one teaching bubble. Since the steps won't be shown in this case, I want to reverse the direction so that the button shows on the bottom right of the bubble. Joey gave a good suggestion below of how to do this without using a useEffect,

Reply:
Oh, cool! Thanks for the information!